presto-main/src/main/java/com/facebook/presto/operator/unnest/NullElementFinder.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.unnest;
+
+import com.facebook.presto.spi.block.Block;
+
+/**
+ * This class checks for the presence of a non-null element in {@code source}, and stores its position.
+ * The result is cached with the first invocation of {@link #getNullElementIndex} after the reset. The
+ * cache can be invalidated by invoking {@link #resetCheck}.
+ */
+class NullElementFinder
```

Comment:
This class is very leaky. It seem to be used only in UnnestBlockBuilder. I suggest to move this logic there.